steering_controllers/src/steering_controllers.cpp

```diff
@@ -0,0 +1,489 @@
+      "~/controller_state", rclcpp::SystemDefaultsQoS());
+    controller_state_publisher_ =
+      std::make_unique<ControllerStatePublisher>(controller_s_publisher_);
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(
+      stderr, "Exception thrown during publisher creation at configure stage with message : %s \n",
+      e.what());
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // TODO(anyone): Reserve memory in state publisher depending on the message type
+  controller_state_publisher_->lock();
+  controller_state_publisher_->msg_.header.stamp = get_node()->now();
+  controller_state_publisher_->msg_.header.frame_id = params_.odom_frame_id;
+  controller_state_publisher_->msg_.odom.pose.pose.position.z = 0;
+  controller_state_publisher_->msg_.odom.child_frame_id = params_.base_frame_id;
+  controller_state_publisher_->unlock();
+  auto & covariance_controller = controller_state_publisher_->msg_.odom.twist.covariance;
+  for (size_t index = 0; index < 6; ++index)
+  {
+    // 0, 7, 14, 21, 28, 35
+    const size_t diagonal_index = NUM_DIMENSIONS * index + index;
+    covariance_controller[diagonal_index] = params_.pose_covariance_diagonal[index];
+    covariance_controller[diagonal_index] = params_.twist_covariance_diagonal[index];
+  }
+  RCLCPP_INFO(get_node()->get_logger(), "configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+void SteeringControllers::reference_callback(const std::shared_ptr<ControllerReferenceMsg> msg)
+{
+  // if no timestamp provided use current time for command timestamp
+  if (msg->header.stamp.sec == 0 && msg->header.stamp.nanosec == 0u)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "Timestamp in header is missing, using current time as command timestamp.");
+    msg->header.stamp = get_node()->now();
+  }
+  const auto age_of_last_command = get_node()->now() - msg->header.stamp;
+
+  if (ref_timeout_ == rclcpp::Duration::from_seconds(0) || age_of_last_command <= ref_timeout_)
+  {
+    input_ref_.writeFromNonRT(msg);
+  }
+  else
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(),
+      "Received message has timestamp %.10f older for %.10f which is more then allowed timeout "
+      "(%.4f).",
+      rclcpp::Time(msg->header.stamp).seconds(), age_of_last_command.seconds(),
+      ref_timeout_.seconds());
+  }
+}
+
+void SteeringControllers::reference_callback_unstamped(
+  const std::shared_ptr<geometry_msgs::msg::Twist> msg)
+{
+  auto twist_stamped = *(input_ref_.readFromNonRT());
+  twist_stamped->header.stamp = get_node()->now();
+  // if no timestamp provided use current time for command timestamp
+  if (twist_stamped->header.stamp.sec == 0 && twist_stamped->header.stamp.nanosec == 0u)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "Timestamp in header is missing, using current time as command timestamp.");
+    twist_stamped->header.stamp = get_node()->now();
+  }
```

Comment:
This check does not make sense here, but we should add deprecation notice, and then later update this in the Navigation stack.

```suggestion
  RCLCPP_WARN(
    get_node()->get_logger(),
    "Use of Twist message without stamped is deprecated and it will be removed in ROS 2 J-Turtle version. Use '~/reference' topic with 'geometry_msgs::msg::TwistStamped' message type in the future.");
```

Reply:
Looks like we don't do anything relevant to the controller within this subscriber as `nput_ref_.writeFromNonRT(msg);` is missing. If you agree, I'll keep all the checks as they are so we can easier test controllers and remove the subscriber in the future.